File: python/Palindrome_Linked_List.py
# https://leetcode.com/problems/palindrome-linked-list/description/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

from collections import deque

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return True
        
        deq = deque()
        node = head

        while node is not None:
            deq.append(node.val)
            node = node.next
        
        while len(deq) > 1:
            if deq.popleft() != deq.pop():
                return False

        return True

File: java/leetcode/PalindromeLinkedList.java
package leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        // ArrayList에 옮겨담기
        List<Integer> arrlst = new ArrayList<>();
        ListNode cur = head;
        while (cur != null) {
            arrlst.add(cur.val);
            cur = cur.next;
        }
        // 오버로딩 메서드 호출
        return isPalindrome(arrlst);
    }

    public boolean isPalindrome(List<Integer> lst) {
        // 양끝에서 투포인터로 팰린드롬 여부 확인
        int left = 0;
        int right = lst.size() - 1;
        while (left < right) {
            if (lst.get(left).equals(lst.get(right))) {
                left++;
                right--;
            } else {
                return false;
            }
        }
        return true;
    }
}